web/modules/custom/dpl_related_content/src/Services/RelatedContent.php

```diff
@@ -219,62 +233,122 @@ public function getContent(): array {
 
-    for ($i = 0; $i < $length;) {
-      if ($i >= $this->maxItems) {
+    for ($i = 0; $i < $max_length; $i++) {
+      if (count($content) >= $this->maxItems) {
         break;
       }
 
-      if (isset($events[$i])) {
+      if ($i < $event_length) {
         $content[] = $event_view_builder->view($events[$i], $this->contentViewMode);
-        $i++;
       }
 
-      if ($i >= $this->maxItems) {
+      if (count($content) >= $this->maxItems) {
         break;
       }
 
-      if (isset($nodes[$i])) {
+      if ($i < $node_length) {
         $content[] = $node_view_builder->view($nodes[$i], $this->contentViewMode);
-        $i++;
       }
     }
 
-    return [
-      '#theme' => 'dpl_related_content',
-      '#title' => $this->title,
-      '#items' => $content,
-      '#list_style' => $this->listStyle,
-      // The results should be cached, but, they have a lot of dependencies -
-      // even depending on the current time (finding future events).
-      // The individual pieces of content are cached themselves, so for the full
-      // result list, we'll add an easily-invalidated cache, with a bunch of
-      // cache tags, and even a time-based cache.
-      '#cache' => [
-        // Max age of 12 hours.
-        '#max-age' => (60 * 60 * 12),
-        '#tags' => ['eventinstance_list', 'eventseries_list', 'node_list'],
-      ],
-    ];
+    return $content;
+  }
+
+  /**
+   * Getting a combined condition group, taking into account AND/OR logic.
+   *
+   * @param \Drupal\Core\Entity\Query\QueryInterface|SelectInterface $query
+   *   The query on which we should create the condition group.
+   * @param array<mixed> $filters
+   *   Array of field_keys => value, that we want to look up.
+   */
+  private function addFilterConditions(QueryInterface|SelectInterface $query, array $filters): void {
+    // Get rid of any empty filters.
+    $filters = array_filter($filters);
+
+    if (empty($filters)) {
+      return;
+    }
+
+    // If we're creating an OR condition, we need to add the filters in their
+    // own condition group, to not pollute other values, such as UUID check.
+    // If it's an AND condition, it doesn't matter, as all filters will be AND.
+    $or_group = $this->andConditions ? NULL : $query->orConditionGroup();
+
+    foreach ($filters as $field_name => $values) {
+      // Get rid of any empty values.
+      $values = array_filter($values);
+
+      if ($or_group) {
+        $or_group->condition($field_name, $values, 'IN');
+        continue;
+      }
+
+      // Looping through, and adding the values.
+      // The reason we add a condition group for every single value, is that
+      // it is the only way it creates JOIN for each value, and allows us to
+      // make "CONTAINS ALL OF X".
+      foreach ($values as $value) {
+        $query->condition($query->andConditionGroup()->condition($field_name, $value));
+      }
+    }
+
+    if ($or_group) {
+      $query->condition($or_group);
+    }
```

Comment:
I do not understand this logic.

I would expect it to be sufficient to:

1. Create an overall condition group of the appropriate type from the query - `and` or `or`
2. For each field create a condition group of the appropriate type - `and` or `or`
3. For each value add a condition to the field condition group
4. Add the field condition to the overall condition

Reply:
I completely agree that it is bonkers that this is necessary.

I also thought what you just mentioned would be the right way forward, but I just couldnt get it to work.
I tried a million things, before [I finally ended up getting help from a colleague in Zulip
](https://reload.zulipchat.com/#narrow/stream/235451-dev-snak/topic/Drupal.2FSymfony.20conditions.20-.20CONTAINS)

That's also why i added all the example content - cause i need to test a trillion times.